src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,671 @@
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	size_t load_offset = (size_t)((void *)ext_lib->desc[lib_id]);
+	void *va_base_text = (void *)mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
+	void *src_txt = (void *)(mod->segment[SOF_MAN_SEGMENT_TEXT].file_offset + load_offset);
+	size_t st_text_size = mod->segment[SOF_MAN_SEGMENT_TEXT].flags.r.length;
+	void *va_base_rodata = (void *)mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
+	void *src_rodata =
+		(void *)(mod->segment[SOF_MAN_SEGMENT_RODATA].file_offset + load_offset);
+	size_t st_rodata_size = mod->segment[SOF_MAN_SEGMENT_RODATA].flags.r.length;
+	int ret;
+
+	st_text_size = st_text_size * CONFIG_MM_DRV_PAGE_SIZE;
+	st_rodata_size = st_rodata_size * CONFIG_MM_DRV_PAGE_SIZE;
+
+	/* Copy Code */
+	ret = lib_manager_load_data_from_storage(va_base_text, src_txt,
+						 st_text_size, SYS_MM_MEM_PERM_EXEC);
+	if (ret < 0) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	/* Copy RODATA */
+	ret = lib_manager_load_data_from_storage(va_base_rodata, src_rodata,
+						 st_rodata_size, SYS_MM_MEM_PERM_EXEC);
+	if (ret < 0) {
+		ret = -ENOMEM;
+		goto err;
+	}
+	/* There are modules marked as lib_code. This is code shared between several modules inside
+	 * the library. Load all lib_code modules with first none lib_code module load.
+	 */
+	if (!mod->type.lib_code)
+		ext_lib->mods_exec_load_cnt++;
+
+	if (ext_lib->mods_exec_load_cnt == 1) {
+		struct sof_man_module *module_entry =
+		    (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(0));
+		for (size_t idx = 0; idx < desc->header.num_module_entries;
+				++idx, ++module_entry) {
+			if (module_entry->type.lib_code) {
+				ret = lib_manager_load_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
+							      idx, mod, desc);
+				if (ret < 0)
+					goto err;
+			}
+		}
+	}
+
+	return 0;
+
+err:
+	sys_mm_drv_unmap_region(va_base_text, st_text_size);
+	sys_mm_drv_unmap_region(va_base_rodata, st_rodata_size);
+
+	return ret;
+}
+
+static int lib_manager_unload_module(uint32_t module_id, struct sof_man_module *mod,
+				     struct sof_man_fw_desc *desc)
+{
+	struct ext_library *ext_lib = ext_lib_get();
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	void *va_base_text = (void *)mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
+	size_t st_text_size = mod->segment[SOF_MAN_SEGMENT_TEXT].flags.r.length;
+	void *va_base_rodata = (void *)mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
+	size_t st_rodata_size = mod->segment[SOF_MAN_SEGMENT_RODATA].flags.r.length;
+	int ret_tmp, ret = 0;
+
+	st_text_size = st_text_size * CONFIG_MM_DRV_PAGE_SIZE;
+	st_rodata_size = st_rodata_size * CONFIG_MM_DRV_PAGE_SIZE;
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_text, st_text_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
+
+	dcache_invalidate_region(va_base_text, st_text_size);
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_rodata, st_rodata_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
+
+	dcache_invalidate_region(va_base_rodata, st_rodata_size);
```

Comment:
wouldn't the mm layer handle cache synchronisation? But in fact - this only synchronises cache on the current core. Couldn't other cores also access that memory? If yes - how do we handle their caches?